src/realm/sync/noinst/client_impl_base.hpp

```diff
@@ -1470,6 +1467,10 @@ inline void ClientImpl::Session::connection_established(bool fast_reconnect)
         ++m_target_download_mark;
     }
 
+    // Call SessionResumed before sending the BIND Message to
```

Comment:
It is primarily here so it doesn't get moved. The role change test needs to be notified of the session resumed/connected prior to sending the BIND message so it can queue up the test command to be sent before the IDENT message. If this notification happens after the BIND message, there isn't enough time for the test to queue up the test command before the IDENT message is sent.

I updated the message to hopefully be more clear.

Reply:
I see. I find it a bit hack-ish though. IIUC, there is a scheduling problem because `send_test_command` posts to the event loop. What if instead we update `m_pending_test_commands` directly under a lock? And then you can invoke it from the event loop when sending BIND (you actually don't need the lock if you create a new method only to be invoked from the event loop). Would that work for your tests? @jbreams what do you think of this approach?

Reply:
I agree, but there isn't a good place to do this, unless I put plumbing in the sync session to add directly to the list of test commands. This is the only thing available in the event hook callback functions.

The current approach ensures the callback to add the test command gets `post`ed before the BIND message is sent and the callback for `async_write_binary()` is run on the event loop to send the next message after the BIND.

Reply:
Yes - the list of test commands sticks around for the lifetime of the `ClientImpl::Session` and the only time commands are removed are when a response is received or the session object is destroyed.
The current `send_test_command()` logic checks to make sure the session is currently active, but that is easy enough to update if needed.

Reply:
the session only becomes inactive if you `pause()` it or destroy the realm - i think in the tests we have right now it should survive a disconnect/reconnect.